### PR TITLE
Fixes to health checks, add a check interval

### PR DESCRIPTION
**Pull request by pimeys**
This was a long stretch and I still don't know did I really do it right. But it now solves an issue in Prisma...

So, first the easy things, this
- adds a health check interval parameter, that limits the health checks to happen a bit less often and we don't need to do a roundtrip to the database with every query.

Then the hard parts:

I've been trying to create a scenario today where I open connection to the database, check that it works, then turn off the database and see what happens in mobc (health checks on). What I noticed is:

- We'll try to open a new connection when the health check fails.
- Another request opens yet another connection, increasing the number of connections counter with one.
- Repeating this we hit the connection limit, and from there on `get` just halts and never returns
- Even opening the database back up the pool is in a halted state and the whole system needs to be restated.

I finally figured out to prevent it from doing that, but I don't know did I break something else with my approach. Open for comments!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -563,25 +579,24 @@ impl<M: Manager> Pool<M> {
             }
         }
 
-        internals.num_open += 1;
```
Comment (importcjj) [inline src/lib.rs]: I think it was a bug there! There `num_open` became incorrect when the `connect` action below was failed.

File: src/lib.rs [hunk 2]
```diff
@@ -563,25 +579,24 @@ impl<M: Manager> Pool<M> {
...
                     created_at: Instant::now(),
                     last_used_at: Instant::now(),
+                    last_checked_at: Instant::now(),
                     brand_new: true,
                 };
 
                 return Ok(conn);
             }
             Err(e) => {
-                let internals = self.0.internals.lock().await;
```
Comment (importcjj) [inline src/lib.rs]: Maybe we should not remove them, cause the `num_open` leak was fixed at line [585](https://github.com/importcjj/mobc/pull/24/files#diff-b4aea3e418ccdb71239b96952d9cddb6R585).

Comment (pimeys) [reply src/lib.rs]: What does that `maybe_open_new_connection` really do here? Will it open a new connection and increment the counter if successful?

Comment (importcjj) [reply src/lib.rs]: When an inner get request fails, a retry will be made, and I expect to have an existing connection available at that time.

Comment (pimeys) [reply src/lib.rs]: I'm adding these two lines here and now the `test_lazy_initialization_failure` never finishes...

Comment (pimeys) [reply src/lib.rs]: Ah, of course it was the Mutex...

**Comment by pimeys** [comment]
Ok this should be ready to merge!

**Pull request merged**